### ISSUE TITLE
Hanging layout/redraw loop in class-book-adjusted

### ISSUE THREAD

**Issue by jasonrclark** [shoes/shoes4#982]
If I run `samples/class-book-adjusted.rb` and drag it down to the narrowest possible width, the app hangs and gives me a beachball (clear sign that it's in some type of infinite loop).

Was happening as far back as `pre2`, so I'm not going to block `pre3` for it, but it's on the short list after that.

Not sure whether it matters or not yet, but I'm running on OS X.

**Comment by jasonrclark** [comment]
Wow, this ended up being weirder than I originally anticipated! Don't have a PR pulled together yet, but thought I'd document what I found and is underway before heading to :zzz: 

Simple repro app:

```ruby
Shoes.app title: "Different Title" do

  stack margin_left: 100 do
    background red
    para "red"
  end
```

The thing to note is that when you squish it horizontally to the point where the red slot gets forced to 0, that's when we get the hang.

![image](https://cloud.githubusercontent.com/assets/130504/5625455/cce9d366-952f-11e4-830f-2721bf464d9a.png)

![image](https://cloud.githubusercontent.com/assets/130504/5625456/d037e1b6-952f-11e4-8245-41df965564d8.png)

... **but** it wasn't actually a resize loop as originally suspected! Instead, the problem was that during the layout for resize, `contents_alignment` for the text block was raising an exception. This was swallowed quietly by SWT, which proceeded to hang! Apparently there are touchy moments in SWT's processing and this is one of them.

Much like we did with the [common `Painter`](https://github.com/shoes/shoes4/blob/779a2490f783b2cdbffb515efa0b4a72eb040902/shoes-swt/lib/shoes/swt/common/painter.rb#L24-L29), I have plans to put a fail-safe trap in the `Slot` class so it will capture any of these failures and keep us from hanging. I will also apply specific fixes at a number of places in the SWT text block business so it doesn't fail under these circumstances (since squeezing it too small should be allowed and shouldn't error like that!)

FWIW, I don't plan to land this for `pre3` since I'm itching to get that out once #1014 is clear, but it should come in the near future after that.
